### ISSUE TITLE
Use project slug for generating name of zipfiles, instead of project title

### ISSUE THREAD

**Issue by tompollard** [MIT-LCP/physionet-build#1674]
We generate a zipfile containing all files within a project. At the moment, it looks like the filename of this zipfile is based on the project title. It would make much more sense for the filename to be based on the project slug, I think. 

The slug is unique, it doesn't have characters like spaces, apostrophes, etc, and it is short. Titles are much less suited to a filename. The function that needs a rethink is copied below:

https://github.com/MIT-LCP/physionet-build/blob/141d2a9262e6b9191e3bf81f3ccee9485c43e9cf/physionet-django/project/modelcomponents/publishedproject.py#L112-L126

**Comment by superryeti** [comment]
Hi @tompollard , I had one concern about this.

Wouldn't changing the names of zipfile mess up the existing zips that were already created?

After checking the codebase, my understanding is that there are several parts of code that initiate zip of project files(which uses the `zip_name` function and create zip of the project files, and there are other part in code that call this `zip_name` function again when someone tries to retrieve the zip file for the project.

So that means when we update this function, the retrieval of zip will break for a lot of existing projects.

I assume to solve that, we need to retrigger the zip of project files, and that might take a lot of resources depending on the number of projects and the file size.. Plus i am not sure if we keep a entry in database, if the project files have been zipped

**Comment by tompollard** [comment]
> Wouldn't changing the names of zipfile mess up the existing zips that were already created?

Yes, this is true. Part of the work would be fixing the current filenames (on our servers, and on the cloud servers). I don't think this would be a huge effort, and if we think it is necessary then it would be better to do sooner than later.

**Comment by superryeti** [comment]
Good point. Do you think somekind of management command might be a good idea to do that? 
If feels weird to have a management command for a one off thing. (but i can't think of anything else right now)

any suggestions?

**Comment by tompollard** [comment]
> Good point. Do you think somekind of management command might be a good idea to do that?
If feels weird to have a management command for a one off thing. (but i can't think of anything else right now). any suggestions?

@bemoody probably has thoughts on this, but I think a management command for the migration would be fine (and it could always be removed later).

**Comment by bemoody** [comment]
Could be a data migration (RunPython) that creates a hard link if the
old zip file already exists.  Or we could just do it by hand.  Don't
want to delete the old files (at least not immediately) because
clients are in the process of downloading them.

**Comment by bemoody** [comment]
Note this problem just came up again (vindr-cxr 1.0.0).

**Comment by bemoody** [comment]
It's uglier than I thought.  The title isn't just used to *name* the zip file, it's also used as the path *inside* the zip file ("enclosing_folder", see LocalProjectFiles.make_zip).

So maybe better to regenerate the zip files as well as renaming them.

**Comment by tompollard** [comment]
https://github.com/MIT-LCP/physionet-build/pull/2134 updates `zip_name` to use more sensible filenames, but includes a `legacy` argument that defaults to the old logic. Can we now:

- Set the default argument to `legacy=False` (or remove it altogether?)
- Update all references to the zipfiles to a try/except (`try: new_zipfile; except legacy_zipfile`)?

**Comment by tompollard** [comment]
@bemoody I think the following management command does what we need to regenerate the zipfiles. 

```
from django.core.management.base import BaseCommand
from project.models import PublishedProject
from physionet.utility import zip_dir
import os


class Command(BaseCommand):
    help = "Regenerates zip files for all published projects using slug-based filenames."

    def handle(self, *args, **options):

        for project in PublishedProject.objects.all():
            slug = project.slug
            version = project.version
            enclosing_folder = f"{slug}-{version}"
            zip_name = os.path.join(project.project_file_root(), f"{enclosing_folder}.zip")

            if os.path.exists(zip_name):
                self.stdout.write(f"Already exists: {zip_name}")
                continue

            self.stdout.write(f"Creating zip for {slug} v{version}")
            self.stdout.write(f"Path: {zip_name}")
            zip_dir(zip_name, target_dir=project.file_root(), enclosing_folder=enclosing_folder)

            project.compressed_storage_size = os.path.getsize(zip_name)
            project.save(update_fields=['compressed_storage_size'])

        self.stdout.write("Done regenerating zips.")
```

This code should be put in a file at `project/management/commands/zipfiles.py`, then run with `python manage.py zipfiles `

After generating the files, we can: 

- adopt the new filename for all zips
- later, delete the old zipfiles

Sound okay?

**Comment by bemoody** [comment]
Thanks for this, and it sounds like a good idea.

However, we don't have disk space to create a zip file of every published project.  At the moment, we don't even have enough disk space to create a second copy of all the zip files we have.

Zip files are useful and I completely understand that *not* having zip files causes its own problems, but for these mega-projects, it's not sustainable to pulblish another entire copy of the data for every new version.

For right now I would say:

(a) Skip projects with compressed_storage_size=0.  Don't create zip files for projects that don't have them.

(b) We should remove, at least, the 5 largest zip files (the ones over 500 GB.)
